Implement Deglob as a CodeAction

jonasbb opened the pull request:
Provide code actions for deglobbing import for wildcard imports, except
for `prelude::*`. To avoid false positives (*) in the code actions, the
deglob command is run before the code action is presented to the user.

The original custom LSP extension is still in place in works. As such
the `Deglob::handle` function is split into a new
`calculate_workspace_edit`, which calculates the data for the deglob
action and is not used to the code action. The original `Deglob::handle`
just calls this function and is reponsible for sending the messages to
the client.

Addresses https://github.com/rust-lang-nursery/rls/issues/515

(*) The path in `CodeAction::handle` only checks for the existence of `use ` and `::*`. False positives could be multiline strings or doc comments.

nrc commented on src/actions/requests.rs:
```diff
@@ -618,37 +636,54 @@ impl<'a> Action<'a> for CodeAction {
 impl<'a> RequestAction<'a> for CodeAction {
     type Response = Vec<Command>;
-    fn handle<O: Output>(&mut self, _id: usize, params: Self::Params, ctx: &mut ActionContext, _out: O) -> Result<Self::Response, ()> {
+    fn handle<O: Output>(&mut self, _id: usize, params: Self::Params, ctx_: &mut ActionContext, _out: O) -> Result<Self::Response, ()> {
```
Could we move the call to `inited` out of `calculate_workspace_edit` so we don't need to keep `ctx_` around?

nrc commented on src/actions/requests.rs:
```diff
@@ -618,37 +636,54 @@ impl<'a> Action<'a> for CodeAction {
+
+        // search for compiler suggestions
+        if let Some(diagnostics) = ctx.previous_build_results.lock().unwrap().get(&file_path) {
+            let suggestions = diagnostics.iter().filter(|&&(ref d, _)| d.range == params.range).flat_map(|&(_, ref ss)| ss.iter());
```
Could you factor this code (and the deglob code) into its own function, so that the handle function is only responsible for deciding what suggestions respond with and compiling them into a result, but not actually making the suggestions?

jonasbb replied:
Sounds good

nrc commented:
Looks good to me, but also r? @Xanewok

Xanewok commented on src/actions/requests.rs:
```diff
@@ -456,22 +462,12 @@ impl<'a> RequestAction<'a> for Rename {
-    type Response = Ack;
-    fn handle<O: Output>(&mut self, id: usize, location: Self::Params, ctx: &mut ActionContext, out: O) -> Result<Self::Response, ()> {
+impl Deglob {
+    fn calculate_workspace_edit(&mut self, location: &<Self as Action>::Params, ctx: &ActionContext) -> Result<DeglobResult, (ErrorCode, &'static str)> {
```
I believe the `&mut self` is unnecessary here

jonasbb replied:
Yes, copy and paste error from handle

Xanewok commented on src/actions/requests.rs:
```diff
@@ -482,21 +478,18 @@ impl<'a> RequestAction<'a> for Deglob {
             let line = match vfs.load_line(&span.file, span.range.row_start) {
```
Can we use the above `map_err(|_| (..))?` approach as above to be consistent?

Xanewok commented on src/actions/requests.rs:
```diff
@@ -618,37 +636,54 @@ impl<'a> Action<'a> for CodeAction {
+            if line.contains("use ") && line.contains("::*") && !line.contains("prelude::*") {
+                let location = Location::new(params.text_document.uri.clone(), params.range);
+                // call Deglob implementation in order to avoid any false positive of the `contains()`
+                if let Ok(deglob) = Deglob.calculate_workspace_edit(&location, ctx_) {
```
Since we're constructing JSON-RPC error response in `Deglob::calculate_workspace_edit`, shouldn't we also do `out.failure_message()` on `Err` here?

jonasbb replied:
No, we shouldn't send failures here. The whole point of this call is to see if it succeeds or not. Consider something like

```
/// ```rust
/// use std::mem::*;
/// ```
```

It contains use and ::* so it is a candidate but since it is just a doc comment rls wouldn't be able to replace anything in here. Therefore, I think it should just ignore errors.

Xanewok replied:
Ah, yes, my bad. I thought the `Err` is unused, but It's currently used [here](https://github.com/rust-lang-nursery/rls/pull/567/files#diff-3761672bafbe5e9917eb056ea9a482e5R576) instead.

Xanewok commented on src/actions/requests.rs:
```diff
@@ -538,27 +527,56 @@ impl<'a> RequestAction<'a> for Deglob {
+    }
+}
+
+impl<'a> RequestAction<'a> for Deglob {
```
To be fair, I would just delete the `rustWorkspace/deglob` extension on RLS side completely.
It's custom and there's a very good chance that only rls-vscode extension implements it, whereas when we switch to using LSP code action and commands we actually improve the support. This is not a very often used feature, I'd say getting and using lightbulb suggestions is more ergonomic and we can remove the logic on the client side fairly easily, so I wouldn't worry about backwards-incompatibility, if that's the case here.

Xanewok commented:
As a side note, I do believe we could loosen some of the assumptions regarding glob imports in general. Since the structure is regular (if I'm not missing anything), we could use a regular expression to parse it and then for matched range we can query the analysis for a deglob. I tried to hack up something quick and came up with something like: `use(\ )*(::)*(\ )*(\ *\w+\ *::)+(\ )*\*(\ )*;`, which roughly matches `use`, optional `::` followed by at least one `scope::` and ends with a glob import `*;`.

Handling multiline imports would probably need more changes and would be more trouble than it's worth, but with this approach we could easily also support deglobbing multiple imports in the same line.

jonasbb commented:
@Xanewok regarding deleting the custom deliberate extensions: In this PR I deliberately exclude prelude wildcard imports, e.g., `std::io::prelude::*`. If someone would like to deglob those the LSP extension is still needed. Is this a problem?

I can try to implement a codeaction for the deglob (currently I'm reusing the applySuggestion one).

jonasbb commented:
@Xanewok @nrc I addressed your comments. Thanks for the feedback. I did not yet delete the custom deglob extension, as it is necessary for `prelude::*;` imports right now. I will change this as well, if you prefer to also show the CodeActions for prelude imports.

Xanewok commented on src/actions/requests.rs:
```diff
@@ -647,27 +649,18 @@ impl ExecuteCommand {
-
+impl CodeAction {
+    /// Create CodeActions for fixes suggested by the compiler
+    /// the results are appended to `code_actions_result`
```
I think it would be a little cleaner to return the actual value from function and `flat_map`/`collect` in CodeAction::handle, but that's just a matter of preference

Xanewok commented on src/actions/requests.rs:
```diff
@@ -477,24 +478,24 @@ impl<'a> RequestAction<'a> for Rename {
 impl Deglob {
-    fn calculate_workspace_edit(&mut self, location: &<Self as Action>::Params, ctx: &ActionContext) -> Result<DeglobResult, (ErrorCode, &'static str)> {
+    fn calculate_workspace_edit(
+        &self,
```
I meant that this doesn't need `&self` at all, same with `make_..._actions()`, but I wouldn't want to block this on that

Xanewok commented:
@jonasbb I'm personally fine with changing default behaviour to not offer deglobs for `prelude::*` imports if that meant removing the custom extension in the server and relying on code action only

jonasbb commented:
@Xanewok Could you please review this again?

I ripped out all code related to the old Deglob struct. This simplified the error handling in the code which finds the globs quite a bit. It will now show a "Deglob Import" for the line containing the wildcard. The line can contain arbitrary many wildcards, which will all be deglobbed at the same time. Even a line like

```rust
use std::io::prelude
::*; // <-
```
this will work now.

I also removed the filter for `prelude::*;` since it will not work in this setting anyway.

Xanewok commented:
@jonasbb 
The approach looks good to me, I'd land this and we could polish it later, if needed.
Could you rebase this, please?
It'd be great if you could also add a deglob test using the new code action method!

nrc commented:
I'm a bit unclear about the prelude imports issue. Could someone explain why they are a special case please?

@Xanewok looks like this has been rebased now

jonasbb commented:
@nrc Preludes are no longer special in the code. My original idea was just that preludes are designed to be used with wildcard imports, thus suggesting a deglob would not be the right thing. It ended up being much easier to ignore this special case though.

Xanewok commented:
@jonasbb if I understand correctly you're writing a test case for that (which required #579)? Do you want me to wait for it here for completeness or would you like to submit that separately?

jonasbb commented:
@Xanewok Yes, I wrote a test case. I added it to this PR since it is still open.

Xanewok commented:
Looks great, thank you for looking into this!
